### DISCUSSION TITLE
Nickel spec libraries and custom output types

### DISCUSSION THREAD

**Discussion by dpc** [tweag/nickel#1304] [Q&A]
Now that v1.0.0 is released, if I was to use it ...to generate:

* Kubernetes - I could just write ad-hoc schemas for stuff I need and then generate yaml/json. Are there any plans or existing project to generate schemas from openapi defintions? (Like e.g. https://github.com/Arnavion/k8s-openapi does for Rust)?
* Terraform - since Terraform uses an HCL syntax - how would one go about generating it? Is some json to hcl transpiler neccessary? (+ Possibly a Nickel schemas for all/most common types. Though I'm not sure if there's anything to generate it from automatically. If not, it could be community maintained human effort, I guess). Anything like that in plans / already being worked on?

**Comment by matthew-healy** [comment]
I'm not aware of any existing work to generate Nickel code from openapi definitions. Sounds like it could be interesting though!

For your second question, @vkleen has been working on [tf-ncl](https://github.com/tweag/tf-ncl) which I think does what you want. I'm not sure it's production ready yet, but I think it's close.

**Reply by vkleen** [reply]
On the point of tf-ncl, Terraform is perfectly to read a configuration as JSON from `main.tf.json`. That's sort of document [here](https://developer.hashicorp.com/terraform/language/syntax/json). That format can be targeted by Nickel pretty easily and tf-ncl is a contract generator to hopefully make it less of a hassle.

**Reply by yannham** [reply]
There had been, some time ago, an aborted tentative of doing that for Kubernetes, including generating contracts from openAPI definitions. It had been left unfinished, and Nickel has changed a lot since then, so I don't know if it'll be of any help, but just in case, here is the subpart of repo about openAPI generation: https://github.com/tweag/nickel-kubernetes/tree/master/openapi2-to-nickel

That being said, tf-ncl is a good inspiration in any case, because even if it doesn't ingest openAPI definitions, it does automatically generate Nickel contracts from schema specifications, which I expect is very similar to what you'd like to do.

One thing to be aware is that "or" clauses (unions) might be problematic to generate automatically, at least for the time being. Cf a similar discussion on JSON Schemas for more context: https://github.com/tweag/nickel/issues/418
